packages/prover/src/utils/consensus.ts

```diff
@@ -0,0 +1,87 @@
+import {Api} from "@lodestar/api/beacon";
+import {allForks, Bytes32, capella} from "@lodestar/types";
+import {GenesisData, Lightclient} from "@lodestar/light-client";
+import {ApiError} from "@lodestar/api";
+import {MAX_PAYLOAD_HISTORY} from "../constants.js";
+import {hexToBuffer} from "./conversion.js";
+
+export async function fetchNearestBlock(
+  api: Api,
+  slot: number,
+  direction: "up" | "down" = "down"
+): Promise<capella.SignedBeaconBlock> {
+  const res = await api.beacon.getBlockV2(slot);
+
+  if (res.ok) return res.response.data;
+
+  if (!res.ok && res.error.code === 404) {
+    return fetchNearestBlock(api, direction === "down" ? slot - 1 : slot + 1);
+  }
+
+  throw new Error(`Can not fetch nearest block for slot=${slot}`);
+}
+
+export async function getSlotLimitsForPayloads(lightClient: Lightclient): Promise<{start: number; end: number}> {
+  const headSlot = lightClient.getHead().beacon.slot;
+  const finalizeSlot = lightClient.getFinalized().beacon.slot;
+  const endSlot = headSlot - MAX_PAYLOAD_HISTORY;
+
+  return {
+    start: headSlot,
+    end: endSlot < finalizeSlot ? finalizeSlot : endSlot,
+  };
+}
+
+export async function getExecutionPayloads(
+  api: Api,
+  startSlot: number,
+  endSlot: number
+): Promise<Record<number, allForks.ExecutionPayload>> {
+  [startSlot, endSlot] = [Math.min(startSlot, endSlot), Math.max(startSlot, endSlot)];
+
+  const payloads: Record<number, allForks.ExecutionPayload> = {};
+
+  let slot = endSlot;
+  let block = (await fetchNearestBlock(api, slot, "down")) as capella.SignedBeaconBlock;
+  payloads[block.message.slot] = block.message.body.executionPayload;
+
+  while (slot >= startSlot) {
+    const previousBlock = (await fetchNearestBlock(api, block.message.slot - 1, "down")) as capella.SignedBeaconBlock;
+
+    if (block.message.body.executionPayload.parentHash === previousBlock.message.body.executionPayload.blockHash) {
+      payloads[block.message.slot] = block.message.body.executionPayload;
```

Comment:
this will easily break on reorgs, just store unfinalized payloads referenced by hash

only condition you need to check in the end if if finalized is ancesor of the head

Reply:
that's just an extra safety check while fetching the blocks from the API. It can be skipped. The re-org scenario is covered on different place.

Reply:
yes just skip the check as anyway this mapping is not canonical since it unfinalized and will cause intractable logical errors